File: c/palindrome_checker/palindrome_checker_5.c
#include <stdio.h>
#include <string.h>
#include <stdbool.h>

// This program checks if a given string is a palindrome.
// A palindrome is a word, phrase, number, or other sequence of characters 
// that reads the same forward and backward (ignoring spaces, punctuation, and capitalization).
// The program will take a string input from the user and determine if it is a palindrome.

bool isPalindrome(char str[]) {
    int len = strlen(str); // Get the length of the string
    int start = 0; // Initialize start index
    int end = len; // Initialize end index

    // Loop through the string from both ends towards the center
    while (start < end) {
        // If characters at start and end indexes are not the same, return false
        if (str[start] != str[end]) {
            return false;
        }
        start++; // Move start index forward
        end--; // Move end index backward
    }
    return true; // If all characters match, return true
}

int main() {
    char str[100]; // Declare a string to hold the input

    // Prompt the user to enter a string
    printf("Enter a string: ");
    fgets(str, sizeof(str), stdin); // Read the input string

    // Remove the newline character from the input string
    str[strcspn(str, "\n")] = 0;

    // Check if the input string is a palindrome
    if (isPalindrome(str)) {
        printf("The string is a palindrome.\n");
    } else {
        printf("The string is not a palindrome.\n");
    }

    return 0; // Return 0 to indicate successful execution
}

File: vbnet/palindrome_checker/palindrome_checker_5.vb
' This program is a palindrome checker written in Visual Basic .NET.
' A palindrome is a word, phrase, number, or other sequence of characters that reads the same forward and backward (ignoring spaces, punctuation, and capitalization).
' The purpose of this program is to check if a given string is a palindrome.
' The program will take input from the user and then determine if the input is a palindrome.

Module PalindromeChecker

    Sub Main()
        ' Prompt the user to enter a string
        Console.WriteLine("Enter a string to check if it is a palindrome:")
        Dim input As String = Console.ReadLine()

        ' Remove spaces and convert to lowercase for uniformity
        Dim processedInput As String = RemoveSpacesAndLowercase(input)

        ' Check if the processed input is a palindrome
        If IsPalindrome(processedInput) Then
            Console.WriteLine("The string is a palindrome.")
        Else
            Console.WriteLine("The string is not a palindrome.")
        End If

        ' Wait for user input before closing the console window
        Console.WriteLine("Press any key to exit.")
        Console.ReadKey()
    End Sub

    ' This function removes spaces from the input string and converts it to lowercase
    Function RemoveSpacesAndLowercase(ByVal str As String) As String
        Dim result As String = ""
        For Each ch As Char In str
            If Not Char.IsWhiteSpace(ch) Then
                result &= Char.ToLower(ch)
            End If
        Next
        Return result
    End Function

    ' This function checks if a given string is a palindrome
    Function IsPalindrome(ByVal str As String) As Boolean
        Dim length As Integer = str.Length
        For i As Integer = 0 To length - 1
            If str(i) <> str(length - i - 1) Then
                Return False
            End If
        Next
        Return True
    End Function

End Module

